### PR TITLE
fix(iOS) CommandBar updates on navigation #3017

### LINKED ISSUE #3017: Changing a CommandBar changes another CommandBar on iOS
## Current behavior

Changing the Background property on the CommandBar on Page1 while navigating to Page2 changes the CommandBar on Page2.

## Expected behavior

Changing the CommandBar on Page1 doesn't affect the CommandBar of Page2.

## How to reproduce it (as minimally and precisely as possible)

Run [this sample](https://github.com/unoplatform/uno/files/4494953/CommandBarSample.zip) on an iOS device.
Once the app is launched, tap the button.
The button navigates to Page2 and sets the background of the Page1's CommandBar to transparent. Howerver, we can see that Page2's CommandBar also becomes transparent.

## Environment

Nuget Package:
Uno.UI
Package Version(s):
2.1.37
Affected platform(s):

- [x] iOS
- [ ] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [x] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### PR DESCRIPTION
**Pull request by ajpinedam**
GitHub Issue (If applicable): #
closes #3017 

## PR Type

What kind of change does this PR introduce?
- Bugfix

## What is the current behavior?
Changing the Background property on the CommandBar on Page1 while navigating to Page2 changes the CommandBar on Page2.

## What is the new behavior?
Changes on one Page CommandBar will not affect other page's CommandBar

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by carldebilly — APPROVED** [review]

**Comment by carldebilly** [comment]
Is there a way to create an automated test for that?

**Comment by ajpinedam** [comment]
> Is there a way to create an automated test for that?

Good question but honestly I am not really sure if this is possible. The test would be time-sensitive in order to reproduce it.

**Comment by jeromelaban** [comment]
For the tests, does it reproduce inside an uno sample control, manually ?

**Comment by ajpinedam** [comment]
> For the tests, does it reproduce inside an uno sample control, manually ?

To reproduce it we would require 2 pages both with `CommandBar` and do a navigation between the pages

**Comment by davidjohnoliver** [comment]
> > For the tests, does it reproduce inside an uno sample control, manually ?
> 
> To reproduce it we would require 2 pages both with `CommandBar` and do a navigation between the pages

Here's a test that's doing that: https://github.com/unoplatform/uno/blob/master/src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/CommandBarTests/UnoSamples_Tests.NativeCommandBar.cs#L68

**Comment by jeromelaban** [comment]
Looking good :) https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=20912&view=ms.vss-test-web.build-test-results-tab&runId=1194840&resultId=100031&paneView=attachments

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UI/Controls/CommandBarHelper.iOS.cs [hunk 1]
```diff
@@ -101,6 +101,15 @@ public static void PageDidDisappear(UIViewController pageController)
 			}
 		}
 
+		public static void PageWillDisappear(UIViewController pageController)
+		{
+			if (pageController.FindTopCommandBar() is { } topCommandBar)
+			{
+				// Set the native navigation bar to null so it does not render when the page is not visible
+				SetNavigationBar(topCommandBar, null!);
```
Comment (davidjohnoliver) [inline src/Uno.UI/Controls/CommandBarHelper.iOS.cs:109]: `null!` is not a thing you should ever write if you can avoid it. Rather you should change the called signature to `SetNavigationItem(CommandBar commandBar, UIKit.UINavigationItem? navigationItem)`

**Review by davidjohnoliver — APPROVED** [review]

File: src/Uno.UI/Controls/CommandBarHelper.iOS.cs (hunk 1 above)
Comment (ajpinedam) [reply src/Uno.UI/Controls/CommandBarHelper.iOS.cs:109]: > `null!` is not a thing you should ever write if you can avoid it. Rather you should change the called signature to `SetNavigationItem(CommandBar commandBar, UIKit.UINavigationItem? navigationItem)`

 I just followed the implementation of similar methods ex: `PageDidDisappear` as I did not want to do refactoring on code I did not add.

@davidjohnoliver

Comment (ajpinedam) [reply src/Uno.UI/Controls/CommandBarHelper.iOS.cs:109]: Will push another PR doing this refactoring

**Comment by jeromelaban** [comment]
@mergifyio backport release/beta/3.1

**Pull request merged**